pipe/section/src/message.rs

```diff
@@ -160,6 +161,161 @@ impl From<Vec<u8>> for Value {
+        match self {
+            Value::U8(v) => Ok(Value::F32(v.into())),
+            Value::U16(v) => Ok(Value::F32(v.into())),
+            Value::I8(v) => Ok(Value::F32(v.into())),
+            Value::I16(v) => Ok(Value::F32(v.into())),
+            t => Err(TypeCastError {
+                msg: format!("could not convert {:?} to Value::F32", t),
+            }),
+        }
+    }
+
+    pub fn into_f64(self) -> Result<Value, TypeCastError> {
+        match self {
+            Value::U8(v) => Ok(Value::F64(v.into())),
+            Value::U16(v) => Ok(Value::F64(v.into())),
+            Value::U32(v) => Ok(Value::F64(v.into())),
+            Value::I8(v) => Ok(Value::F64(v.into())),
+            Value::I16(v) => Ok(Value::F64(v.into())),
+            Value::I32(v) => Ok(Value::F64(v.into())),
+            Value::F32(v) => Ok(Value::F64(v.into())),
+            t => Err(TypeCastError {
+                msg: format!("could not convert {:?} to Value::F64", t),
+            }),
+        }
+    }
+
+    pub fn into_i16(self) -> Result<Value, TypeCastError> {
+        match self {
+            Value::U8(v) => Ok(Value::I16(v.into())),
+            Value::I8(v) => Ok(Value::I16(v.into())),
+            t => Err(TypeCastError {
+                msg: format!("could not convert {:?} to Value::I16", t),
+            }),
+        }
+    }
+
+    pub fn into_i32(self) -> Result<Value, TypeCastError> {
+        match self {
+            Value::I8(v) => Ok(Value::I32(v.into())),
+            Value::I16(v) => Ok(Value::I32(v.into())),
+            Value::U8(v) => Ok(Value::I32(v.into())),
+            Value::U16(v) => Ok(Value::I32(v.into())),
+            t => Err(TypeCastError {
+                msg: format!("could not convert {:?} to Value::I32", t),
+            }),
+        }
+    }
+
+    pub fn into_i64(self) -> Result<Value, TypeCastError> {
+        match self {
+            Value::I8(v) => Ok(Value::I64(v.into())),
+            Value::I16(v) => Ok(Value::I64(v.into())),
+            Value::I32(v) => Ok(Value::I64(v.into())),
+            Value::U8(v) => Ok(Value::I64(v.into())),
+            Value::U16(v) => Ok(Value::I64(v.into())),
+            Value::U32(v) => Ok(Value::I64(v.into())),
+            t => Err(TypeCastError {
+                msg: format!("could not convert {:?} to Value::I32", t),
+            }),
+        }
+    }
+
+    pub fn into_decimal(self) -> Result<Value, TypeCastError> {
+        match self {
+            Value::I8(v) => Ok(Value::Decimal(v.into())),
+            Value::I16(v) => Ok(Value::Decimal(v.into())),
+            Value::I32(v) => Ok(Value::Decimal(v.into())),
+            Value::I64(v) => Ok(Value::Decimal(v.into())),
+            Value::U8(v) => Ok(Value::Decimal(v.into())),
+            Value::U16(v) => Ok(Value::Decimal(v.into())),
+            Value::U32(v) => Ok(Value::Decimal(v.into())),
+            Value::U64(v) => Ok(Value::Decimal(v.into())),
+            // TODO: figure out how to cast these
```

Comment:
not sure what is the problem here exactly, rust_decimal provides methods to build Decimal from [f32/f64](https://docs.rs/rust_decimal/latest/rust_decimal/prelude/trait.FromPrimitive.html#method.from_f32)

Reply:
So we can return typecasterror from this function, but in the case of f32 and f64 they are only unrepresentable some times. You could have a value that works fine in development but rolls over at some point when it's out in the field. We need a better way to handle errors to make this friendly to use.